clvm-derive/src/to_clvm.rs

```diff
@@ -1,53 +1,230 @@
 use proc_macro2::{Ident, Span, TokenStream};
-use quote::quote;
-use syn::{parse_quote, Data, DeriveInput, Fields, GenericParam, Index, TypeParam};
+use quote::{quote, ToTokens};
+use syn::{
+    parse_quote, spanned::Spanned, Data, DeriveInput, Expr, Fields, FieldsNamed, FieldsUnnamed,
+    GenericParam, Index,
+};
 
-use crate::helpers::{add_trait_bounds, parse_args, Repr};
+use crate::{
+    helpers::{add_trait_bounds, parse_clvm_attr, Repr},
+    macros::{repr_macros, Macros},
+};
 
-pub fn to_clvm(mut ast: DeriveInput) -> TokenStream {
-    let args = parse_args(&ast.attrs);
-    let crate_name = quote!(clvm_traits);
+#[derive(Default)]
+struct FieldInfo {
+    field_names: Vec<Ident>,
+    field_accessors: Vec<TokenStream>,
+    initializer: TokenStream,
+}
+
+struct VariantInfo {
+    name: Ident,
+    value: Expr,
+    field_info: FieldInfo,
+    macros: Macros,
+}
 
-    let field_names: Vec<TokenStream>;
+pub fn to_clvm(ast: DeriveInput) -> TokenStream {
+    let clvm_attr = parse_clvm_attr(&ast.attrs);
+    let crate_name = quote!(clvm_traits);
 
     match &ast.data {
-        Data::Struct(data_struct) => match &data_struct.fields {
-            Fields::Named(fields) => {
-                let fields = &fields.named;
-                field_names = fields
-                    .iter()
-                    .map(|field| {
-                        let ident = field.ident.clone().unwrap();
-                        quote!(#ident)
-                    })
-                    .collect();
+        Data::Struct(data_struct) => {
+            if clvm_attr.untagged {
+                panic!("cannot use `untagged` on a struct");
             }
-            Fields::Unnamed(fields) => {
-                let fields = &fields.unnamed;
-                field_names = fields
-                    .iter()
-                    .enumerate()
-                    .map(|(i, _)| {
-                        let index = Index::from(i);
-                        quote!(#index)
-                    })
-                    .collect();
+            let macros = repr_macros(&crate_name, clvm_attr.expect_repr());
+            let field_info = fields(&data_struct.fields);
+            impl_for_struct(&crate_name, &ast, &macros, &field_info)
+        }
+        Data::Enum(data_enum) => {
+            if !clvm_attr.untagged && clvm_attr.repr == Some(Repr::Curry) {
+                panic!("cannot use `curry` on a tagged enum");
+            }
+
+            let mut next_value: Expr = parse_quote!(0);
```

Comment:
is this the "tag" you will use to convert the enum into CLVM?
I associate the term "tag" with "tagged union", later "variant" in the C++ world. This is of course essentially the same as a rust enum. Is the term "tag" used in rust as well? I get the impression the term "discriminant" is used as the underlying value representing which enum variant is active.

If so, it seems like a better name for this variable would be `next_discriminant` or maybe `next_tag`. But perhaps best would be to have a comment describing what this is used for.

Reply:
Improved the name